### PR TITLE
Update BatchMonteCarloExpectedImprovement

### PR DESCRIPTION
**Pull request by uri-granta**
Update BatchMonteCarloExpectedImprovement acquisition function builder to avoid retracing each time.

### CODE REVIEW

File: tests/integration/test_bayesian_optimization.py [hunk 1]
```diff
@@ -73,7 +73,7 @@
                 ),
             ),
             (
-                10,
+                12,
```
Comment (uri-granta) [inline tests/integration/test_bayesian_optimization.py:76]: This is purely due to the BatchReparametrizationSampler picking different _eps values to before.

File: trieste/acquisition/sampler.py [hunk 2]
```diff
@@ -269,6 +269,10 @@ def __init__(self, sample_size: int, model: ProbabilisticModel):
             tf.ones([0, 0, sample_size], dtype=tf.float64), shape=[None, None, sample_size]
         )  # [0, 0, S]
 
+        # for some reason graph compilation is resulting in self._eps reporting the wrong shape
+        # we therefore use an extra boolean variable to keep track of whether it's initialised
+        self._initialized = tf.Variable(False)
```
Comment (henrymoss) [inline trieste/acquisition/sampler.py:274]: Why does this have to be a tf Variable?

**Review by henrymoss — COMMENTED** [review]
LGTM

File: trieste/acquisition/sampler.py (hunk 2 above)
Comment (uri-granta) [reply trieste/acquisition/sampler.py:274]: For the same reason self._eps needs to be: so it can be updated in the compiled graph.

**Comment by uri-granta** [comment]
> LGTM

You ok to tick then?

**Review by henrymoss — APPROVED** [review]

**Pull request merged**